app/master/slave.py

```diff
@@ -244,6 +250,87 @@ def update_last_heartbeat_time(self):
     def get_last_heartbeat_time(self) -> datetime:
         return self._last_heartbeat_time
 
+    def _update_slave_registry(self):
+        """
+        Remove shutdown-ed slave from SlaveRegistry.
+        """
+        self._logger.info('Removing slave (url={}; id={}) from Slave Registry.'.format(self.url, self.id))
+        SlaveRegistry.singleton().remove_slave(slave_url=self.url)
+
+
+class SlaveRegistry(Singleton):
+    """
+    SlaveRegistry class is a singleton class which stores and maintains list of connected slaves.
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.all_slaves_by_url = {}
+        self.all_slaves_by_id = {}
+        self._slave_dict_lock = Lock()
```

Comment:
I don't think this lock is necessary. If i'm not mistaken SlaveRegistry methods are only being called by ClusterMaster, which is only being called by tornado's thread. You can verify (or falsify) this claim by printing [threading.get_ident()](https://docs.python.org/3/library/threading.html#threading.get_ident) in slave registry methods. See if the identifier ever changes.

Reply:
Make sure to document this assumption if you remove the lock.

Reply:
I do not agree. `remove_slave` is called from `Slave.mark_slave_idle`.